frontera/contrib/backends/redis_backend/__init__.py

```diff
@@ -0,0 +1,376 @@
+# -*- coding: utf-8 -*-
+from collections import Iterable
+from datetime import datetime
+from frontera.utils.url import parse_domain_from_url_fast
+from frontera import DistributedBackend
+from frontera.core.components import Metadata, Queue, States
+from frontera.contrib.backends.partitioners import Crc32NamePartitioner
+from frontera.utils.misc import get_crc32, load_object
+import logging
+from msgpack import packb, unpackb
+from redis import ConnectionPool, StrictRedis
+from time import time
+
+FIELD_CRAWL_AT = b'crawl_at'
+FIELD_CREATED_AT = b'created_at'
+FIELD_DEPTH = b'depth'
+FIELD_DOMAIN = b'domain'
+FIELD_DOMAIN_FINGERPRINT = b'domain_fingerprint'
+FIELD_ERROR = b'error'
+FIELD_FINGERPRINT = b'fingerprint'
+FIELD_NAME = b'name'
+FIELD_SCORE = b'score'
+FIELD_STATE = b'state'
+FIELD_STATUS_CODE = b'status_code'
+FIELD_URL = b'url'
+
+
+class RedisQueue(Queue):
+    MAX_SCORE = 1.0
+    MIN_SCORE = 0.0
+    SCORE_STEP = 0.01
+
+    def __init__(self, manager, pool, partitions, delete_all_keys=False):
+        settings = manager.settings
+        codec_path = settings.get('BACKEND_CODEC')
+        encoder_cls = load_object(codec_path + ".Encoder")
+        decoder_cls = load_object(codec_path + ".Decoder")
+        self._encoder = encoder_cls(manager.request_model)
+        self._decoder = decoder_cls(manager.request_model, manager.response_model)
+        self._pool = pool
+        self._partitions = [i for i in range(0, partitions)]
+        self._partitioner = Crc32NamePartitioner(self._partitions)
+        self._logger = logging.getLogger("redis_backend.queue")
+
+        if delete_all_keys:
+            connection = StrictRedis(connection_pool=self._pool)
+            connection.flushdb()
+
+    def get_next_requests(self, max_n_requests, partition_id, **kwargs):
+        """
+        Fet new batch from priority queue.
+        :param max_n_requests: maximum number of requests
+        :param partition_id: partition id to get batch from
+        :return: list of :class:`Request <frontera.core.models.Request>` objects.
+        """
+        max_requests_per_host = kwargs.pop('max_requests_per_host')
+        connection = StrictRedis(connection_pool=self._pool)
+        queue = {}
+        count = 0
+        now_ts = int(time())
+        max_host_items = 0
+        to_remove = []
+        for data in connection.zrevrange(partition_id, start=0, end=max_n_requests):
+            item = unpackb(data, use_list=False)
+            timestamp, fprint, host_crc32, _, score = item
+            if timestamp > now_ts:
```

Comment:
What if there `max_n_requests` (or more) high-priority requests scheduled to be crawled tomorrow (in a row) followed by sequence of regular mid-priority "crawl now" requests?